egs/librispeech/asr/simple_v1/ctc_att_transformer_train.py

```diff
@@ -0,0 +1,472 @@
+        if att_rate != 0.0:
+            att_loss = model.decoder_forward(encoder_memory, memory_mask, supervisions, graph_compiler)
+    else:
+        with torch.no_grad():
+            nnet_output, encoder_memory, memory_mask = model(feature, supervision_segments)
+            if att_rate != 0.0:
+                att_loss = model.decoder_forward(encoder_memory, memory_mask, supervisions, graph_compiler)
+
+    # nnet_output is [N, C, T]
+    nnet_output = nnet_output.permute(0, 2, 1)  # now nnet_output is [N, T, C]
+
+    decoding_graph = graph_compiler.compile(texts).to(device)
+
+    # nnet_output2 = nnet_output.clone()
+    # blank_bias = -7.0
+    # nnet_output2[:,:,0] += blank_bias
+
+    dense_fsa_vec = k2.DenseFsaVec(nnet_output, supervision_segments)
+    assert decoding_graph.is_cuda()
+    assert decoding_graph.device == device
+    assert nnet_output.device == device
+
+    target_graph = k2.intersect_dense(decoding_graph, dense_fsa_vec, 10.0)
+
+    tot_scores = target_graph.get_tot_scores(
+        log_semiring=True,
+        use_double_scores=True)
+
+    (tot_score, tot_frames,
+     all_frames) = get_tot_objf_and_num_frames(tot_scores,
+                                               supervision_segments[:, 2])
+
+    if is_training:
+        if att_rate != 0.0:
+            loss = ( - (1.0 - att_rate) * tot_score + att_rate * att_loss) / (len(texts) * accum_grad)
+        else:
+            loss = (-tot_score) / (len(texts) * accum_grad)
+        loss.backward()
+        if is_update:
+            clip_grad_value_(model.parameters(), 5.0)
+            optimizer.step()
+            optimizer.zero_grad()
+
+    ans = -tot_score.detach().cpu().item(), tot_frames.cpu().item(
+    ), all_frames.cpu().item()
+    return ans
+
+
+def get_validation_objf(dataloader: torch.utils.data.DataLoader,
+                        model: AcousticModel, device: torch.device,
+                        graph_compiler: CtcTrainingGraphCompiler):
+    total_objf = 0.
+    total_frames = 0.  # for display only
+    total_all_frames = 0.  # all frames including those seqs that failed.
+
+    model.eval()
+
+    for batch_idx, batch in enumerate(dataloader):
+        objf, frames, all_frames = get_objf(
+            batch=batch,
+            model=model,
+            device=device,
+            graph_compiler=graph_compiler,
+            is_training=False,
+            is_update=False, 
+        )
+        total_objf += objf
+        total_frames += frames
+        total_all_frames += all_frames
+
+    return total_objf, total_frames, total_all_frames
+
+
+def train_one_epoch(dataloader: torch.utils.data.DataLoader,
+                    valid_dataloader: torch.utils.data.DataLoader,
+                    model: AcousticModel, device: torch.device,
+                    graph_compiler: CtcTrainingGraphCompiler,
+                    optimizer: torch.optim.Optimizer,
+                    accum_grad: int,
```

Comment:
This function is getting a lot of args, we should probably start documenting it.
What is the motivation for accum_grad, is it to do with multi-GPU training to reduce sync costs?
Because if there's one GPU I don't see what the benefit would be, you could just use a smaller minibatch size and if necessary adjust the learning rate.

Reply:
Will add documents.
`accum_grad` is used with one GPU. Maybe using smaller batch size and adjust learning rate could achieve similar results. But I think it's more convenient to use `accum_grad` to simulate larger batch size. And for users who don't have sufficient GPU memory, it would be easier to achieve similar results. They just need to change the `max_frames` and `accum_grad` and don't need to tuning learning rate.

Reply:
mm OK, but bear in mind that theoretically, it's always best to do the update as soon as possible, so the default recipe shouldn't be using more than 1 frame.  (May have to reduce learning rate.)

Reply:
Sure. Will update it.